kafka/src/main/java/com/metamx/tranquility/kafka/writer/TranquilityEventWriter.java

```diff
@@ -0,0 +1,231 @@
+
+    final Timestamper<Map<String, Object>> timestamper = new Timestamper<Map<String, Object>>()
+    {
+      @Override
+      public DateTime timestamp(Map<String, Object> theMap)
+      {
+        return new DateTime(theMap.get(timestampColumn));
+      }
+    };
+
+    curator.start();
+
+    Beam<Map<String, Object>> druidService = DruidBeams
+        .builder(timestamper)
+        .curator(curator)
+        .discoveryPath(discoveryPath)
+        .location(
+            DruidLocation.create(
+                indexService,
+                firehosePattern,
+                dataSource
+            )
+        )
+        .timestampSpec(schema.getParser().getParseSpec().getTimestampSpec())
+        .rollup(
+            DruidRollup.create(
+                DruidDimensions.specific(getDimensions(schema)).withSpatialDimensions(getSpatialDimensions(schema)),
+                Lists.newArrayList(schema.getAggregators()),
+                schema.getGranularitySpec().getQueryGranularity()
+            )
+        )
+        .tuning(
+            ClusteredBeamTuning
+                .builder()
+                .segmentGranularity(schema.getGranularitySpec().getSegmentGranularity())
+                .windowPeriod(tranquilityConfig.getWindowPeriod())
+                .partitions(tranquilityConfig.getPartitions())
+                .replicants(tranquilityConfig.getReplicants())
+                .build()
+        )
+        .buildBeam();
+
+    BeamPacketizerListener<Map<String, Object>> beamPacketizerListener =
+        new BeamPacketizerListener<Map<String, Object>>()
+    {
+      @Override
+      public void ack(Map<String, Object> map)
+      {
+        listener.onSuccess(map);
+      }
+
+      @Override
+      public void fail(Throwable throwable, Map<String, Object> map)
+      {
+        listener.onFailure(throwable, map);
+      }
+    };
+
+    beamPacketizer = new BeamPacketizer<>(
+        druidService,
+        beamPacketizerListener,
+        tranquilityConfig.getBatchSize(),
+        tranquilityConfig.getMaxPendingBatches()
+    );
+
+    beamPacketizer.start();
+  }
+
+  @Override
+  public synchronized boolean send(String message) throws InterruptedException
+  {
+    if (beamPacketizer == null) {
+      throw new IllegalStateException("Call init() first");
+    }
+
+    Map<String, Object> map;
+    try {
+      map = mapper.readValue(message, new TypeReference<HashMap<String, Object>>() { });
+    } catch (Exception e) {
+      log.error(e, "MESSAGE REJECTED - Failed to parse input data as JSON for: %s", message);
+      return false;
+    }
+
+    try {
+      new DateTime(map.get(timestampColumn));
+    } catch (Exception e) {
+      log.error(e, "MESSAGE REJECTED - Failed to parse timestamp for: %s", map.get(timestampColumn));
+      return false;
+    }
+
+    beamPacketizer.send(map);
+
+    return true;
+  }
+
+  @Override
+  public void flush() throws InterruptedException
```

Comment:
*All* access to a non-thread-safe object needs to be synchronized somehow, since without synchronization points there is no guarantee that different threads will see consistent versions of those data structures ("visibility" on http://gee.cs.oswego.edu/dl/cpj/jmm.html)